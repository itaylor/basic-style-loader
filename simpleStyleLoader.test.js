const path = require('path');
const fs = require('fs');

const rimraf = require('rimraf');
const webpack = require('webpack');
const jsdom = require('jsdom');

const outputDir = `${__dirname}/test-output`;
const fixturesDir = `${__dirname}/fixtures`;

let window;
describe('', () => {
  beforeEach(() => {
    rimraf.sync(outputDir);
    const jsdomHtml = fs.readFileSync(`${fixturesDir}/index.html`);
    const virtualConsole = new jsdom.VirtualConsole();
    virtualConsole.sendTo(console);
    const obj = new jsdom.JSDOM(jsdomHtml, {
      resources: 'usable',
      runScripts: 'dangerously',
      virtualConsole,
    });
    window = obj.window; // eslint-disable-line prefer-destructuring
  });
  afterAll(() => {
    rimraf.sync(outputDir);
  });
 
  test('basic: scripts get added to head', async () => {
    const compiler = createWebpackCompiler(getWebpackConfig());
    await compileAndRun(compiler);
    const resultedStyleTags = window.document.querySelectorAll(['head style']);
    expect(resultedStyleTags.length).toBe(3);
    expect(resultedStyleTags).toMatchSnapshot();
  });
  test('options: sourceMap option adds inline sourceMaps if css-loader passes them', async () => {
    const use = [{
      loader: 'simple-style-loader', 
      options: {
        sourceMap: true,
      }
    }, {
      loader: 'css-loader', 
      options: {
        sourceMap: true,
      }
    }];
    const compiler = createWebpackCompiler(getWebpackConfig(use));
    await compileAndRun(compiler);
    const resultedStyleTags = window.document.querySelectorAll(['head style']);
    expect(resultedStyleTags.length).toBe(3);
    expect(resultedStyleTags).toMatchSnapshot();
  });
  test('options: sourceMap option wont add sourcemap if css-loader does not also', async () => {
    const use = [{
      loader: 'simple-style-loader', 
      options: {
        sourceMap: true,
      }
    }, {
      loader: 'css-loader', 
      options: {
        sourceMap: false,
      }
    }];
    const compiler = createWebpackCompiler(getWebpackConfig(use));
    await compileAndRun(compiler);
    const resultedStyleTags = window.document.querySelectorAll(['head style']);
    expect(resultedStyleTags.length).toBe(3);
    expect(resultedStyleTags).toMatchSnapshot();
  });
  test('options: sourceMap prevents sourcemap even if css-loader passes one', async () => {
    const use = [{
      loader: 'simple-style-loader', 
      options: {
        sourceMap: false,
      }
    }, {
      loader: 'css-loader', 
      options: {
        sourceMap: true,
      }
    }];
    const compiler = createWebpackCompiler(getWebpackConfig(use));
    await compileAndRun(compiler);
    const resultedStyleTags = window.document.querySelectorAll(['head style']);
    expect(resultedStyleTags.length).toBe(3);
    expect(resultedStyleTags).toMatchSnapshot();
  });
  test('options: insertionFn option allows tags to be placed elsewhere in the DOM', async () => {
    const use = [{
      loader: 'simple-style-loader', 
      options: {
        insertionFn: (styleTag) => {
          document.body.appendChild(styleTag);
        },
      }
    }, 
    'css-loader'
    ];
    const compiler = createWebpackCompiler(getWebpackConfig(use));
    await compileAndRun(compiler);
    const resultedStyleTags = window.document.querySelectorAll(['body style']);
    expect(resultedStyleTags.length).toBe(2);
    expect(resultedStyleTags).toMatchSnapshot();
  });
  test('options: insertionFn gets passed styleObj and can use it to add attributes', async () => {
    const use = [{
      loader: 'simple-style-loader', 
      options: {
        insertionFn: (styleTag, styleObj) => { 
          const idSafe = `id-${styleObj.id}`.replace(/[\W]/gi, '_');
          styleTag.setAttribute('id', idSafe);
          document.head.appendChild(styleTag);
        },
      }
    }, 
    'css-loader'
    ];
    const compiler = createWebpackCompiler(getWebpackConfig(use));
    await compileAndRun(compiler);
    const resultedStyleTags = window.document.querySelectorAll(['head style']);
    expect(resultedStyleTags.length).toBe(3);
    expect(resultedStyleTags).toMatchSnapshot();
  });
});

function compileAndRun(compiler, compiledScriptName = 'bundle.js') {
  let resolve;
  let reject;
  const promise = new Promise((res, rej) =>  {
    resolve = res;
    reject = rej;
  });

  compiler.run((err, stats) => {
    if (err) {
      return reject(err);
    }
    if (stats.compilation.errors.length) {
      return reject(new Error(stats.compilation.errors));
    }
    const bundleJs = stats.compilation.assets[compiledScriptName].source();
    window.eval(bundleJs); 
    return resolve();
  });
  return promise;
}

function createWebpackCompiler(webpackConfig) {
  Object.assign(webpackConfig, {
    mode: 'development',
    resolveLoader: {
      alias: {
        'simple-style-loader': path.resolve(__dirname),
      },
    },
    plugins: [],
  });
  const compiler = webpack(webpackConfig);
  return compiler;
}

function getWebpackConfig(use = ['simple-style-loader','css-loader']) {
  const webpackConfig = {
    entry: `${fixturesDir}/main.js`,
    output: {
      path: outputDir,
      filename: 'bundle.js',
    },
    module: {
      rules: [{
        test: /\.css?$/,
        use,
      }],
    },
  };
  return webpackConfig;
} 