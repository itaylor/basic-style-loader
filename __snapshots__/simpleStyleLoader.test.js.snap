// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic: scripts get added to head 1`] = `
NodeList [
  <style
    id="existing-style"
  >
    .existing { color: yellow }
  </style>,
  <style
    type="text/css"
  >
    .required { color: blue }
  </style>,
  <style
    type="text/css"
  >
    .requiredTwo { color: cyan }
  </style>,
]
`;

exports[`options: insertionFn gets passed styleObj and can use it to add attributes 1`] = `
NodeList [
  <style
    id="existing-style"
  >
    .existing { color: yellow }
  </style>,
  <style
    id="id___node_modules_css_loader_dist_cjs_js___fixtures_requiredOne_css"
    type="text/css"
  >
    .required { color: blue }
  </style>,
  <style
    id="id___node_modules_css_loader_dist_cjs_js___fixtures_requiredTwo_css"
    type="text/css"
  >
    .requiredTwo { color: cyan }
  </style>,
]
`;

exports[`options: insertionFn option allows tags to be placed elsewhere in the DOM 1`] = `
NodeList [
  <style
    type="text/css"
  >
    .required { color: blue }
  </style>,
  <style
    type="text/css"
  >
    .requiredTwo { color: cyan }
  </style>,
]
`;

exports[`options: sourceMap option adds inline sourceMaps if css-loader passes them 1`] = `
NodeList [
  <style
    id="existing-style"
  >
    .existing { color: yellow }
  </style>,
  <style
    type="text/css"
  >
    .required { color: blue }
/*# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbInJlcXVpcmVkT25lLmNzcyJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiQUFBQSxZQUFZLFlBQVkiLCJmaWxlIjoicmVxdWlyZWRPbmUuY3NzIiwic291cmNlc0NvbnRlbnQiOlsiLnJlcXVpcmVkIHsgY29sb3I6IGJsdWUgfSJdfQ== */
  </style>,
  <style
    type="text/css"
  >
    .requiredTwo { color: cyan }
/*# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbInJlcXVpcmVkVHdvLmNzcyJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiQUFBQSxlQUFlLFlBQVkiLCJmaWxlIjoicmVxdWlyZWRUd28uY3NzIiwic291cmNlc0NvbnRlbnQiOlsiLnJlcXVpcmVkVHdvIHsgY29sb3I6IGN5YW4gfSJdfQ== */
  </style>,
]
`;

exports[`options: sourceMap option wont add sourcemap if css-loader does not also 1`] = `
NodeList [
  <style
    id="existing-style"
  >
    .existing { color: yellow }
  </style>,
  <style
    type="text/css"
  >
    .required { color: blue }
  </style>,
  <style
    type="text/css"
  >
    .requiredTwo { color: cyan }
  </style>,
]
`;

exports[`options: sourceMap prevents sourcemap even if css-loader passes one 1`] = `
NodeList [
  <style
    id="existing-style"
  >
    .existing { color: yellow }
  </style>,
  <style
    type="text/css"
  >
    .required { color: blue }
  </style>,
  <style
    type="text/css"
  >
    .requiredTwo { color: cyan }
  </style>,
]
`;
